src/serializer.ts

```diff
@@ -94,26 +129,25 @@ export class Serializer {
         }
         //First of all, we'll create an instance of our class
         const result: any = this.getInstance<T>(obj, clazz);
-        //Then we copy every property of our object to our clazz
-        for (const prop in obj) {
-            //Simple check to avoid iterations over strange things.
-            if (obj.hasOwnProperty(prop)) {
-                //We get our metadata for the class to deserialize
-                const metadata: new() => any = Reflect.getMetadata('serializer:class', result, prop);
-                //If we have some metadata, we'll handle them
-                if (metadata !== undefined) {
-                    if (obj[prop] instanceof Array) {
-                        result[prop] = [];
-                        for (const item of obj[prop]) {
-                            result[prop].push(this.deserialize(item, metadata));
-                        }
-                    } else {
-                        result[prop] = this.deserialize(obj[prop], metadata);
+        //And we get the property binding map.
+        const properties = Serializer.getPropertyMap(obj, result);
+        //Then we copy every property of our object to our instance, using bindings.
+        for (const prop in properties) {
+            //We get our metadata for the class to deserialize.
+            const className: new() => any = Reflect.getMetadata(METADATA_DESERIALIZE_AS, result, properties[prop]);
+            //If we have some class-related metadata, we'll handle them.
+            if (className !== undefined) {
+                if (obj[prop] instanceof Array) {
+                    result[properties[prop]] = [];
+                    for (const item of obj[prop]) {
+                        result[properties[prop]].push(this.deserialize(item, className));
```

Comment:
You can still do something like :

```typescript
for (const prop in properties) {
    const className: new() => any = Reflect.getMetadata(METADATA_DESERIALIZE_AS, result, properties[prop]);
    result[properties[prop]] = this.magicalPropertyResolver(obj, prop, className);
}
```

I'm sure we can do it even more sexy, because passing all three parameters looks not that good... but the recursivity is not the problem here